### ISSUE TITLE
switch hdsjdk build system to gradle

### ISSUE THREAD

**Issue by lbergelson** [samtools/htsjdk#377]
I recommend that we switch htsjdk to use gradle as our build system.

Ant is old and not very beautiful.  I know from personal experience and anecdotal experience from those working around me that the ant build isn't particularly well supported by intellij.

The GATK4 team has now gained experience working with gradle and it's been a very pleasant tool to use.  Code for common things is very short and readable, and more complex things are reasonable because it's written in groovy instead of XML.

The following is a simple build.gradle script that lets you pull down dependencies, build htsjdk, and run tests.  A slightly more fully featured version can be seen in my branch https://github.com/samtools/htsjdk/tree/lb_gradle_build.  (It has code to reduce test verbosity on travis, which is something that we've had problems controlling using ant, as well as hooks for coveralls.)

```
buildscript {
    repositories {
        mavenCentral()
    }
}

apply plugin: 'java'
apply plugin: 'idea'

repositories {
    mavenCentral()
}

dependencies {
    compile files("${System.properties['java.home']}/../lib/tools.jar")

    compile "org.apache.commons:commons-jexl:2.1.1"
    compile "commons-logging:commons-logging:1.1.1"
    compile "org.xerial.snappy:snappy-java:1.0.3-rc3"
    compile "org.apache.commons:commons-compress:1.4.1"
    compile "org.tukaani:xz:1.5"
    compile "org.apache.ant:ant:1.8.2"

    testCompile "org.testng:testng:6.8.8"
}

sourceCompatibility = 1.8
targetCompatibility = 1.8

version="2.0.0"

logger.info("build for version:" + version)
group = 'com.github.samtools'

jar {
    manifest {
        attributes 'Implementation-Title': 'HTSJDK',
                'Implementation-Vendor' : 'Broad Institute',
                'Implementation-Version': version
    }
}


test {
    useTestNG{}

    // set heap size for the test JVM(s)
    minHeapSize = "1G"
    maxHeapSize = "2G"
}

task wrapper(type: Wrapper) {
    gradleVersion = '2.8'
}

// This is a hack to disable the java 8 default javadoc lint until we fix the html formatting
if (JavaVersion.current().isJava8Compatible()) {
    tasks.withType(Javadoc) {
        options.addStringOption('Xdoclint:none', '-quiet')
    }
}

```

I think gradle's builds are much more readable and understandable than ant.  Intellij also has excellent gradle support which is helpful.  

Since gradle includes publishing support similar to what the sbt script is doing, it would be easy to remove the sbt script as well, and have only 1 build system in the repository.  This would be a major improvement for users (and IDEs) who get confused by the presence of 2 build files.  

Switching to gradle is easy, since we can package a gradlew wrapper file, a shell script which can be run instead of a gradle installation, which downloads the correct version of gradle an executes it.

The downsides as I see them:
- *People are already used to ant and change is confusing:* This is a real problem, but I think there is minimal friction to change, the commands are similar `ant clean test` is not too hard to replace with `gradle clean test`
- *Picard uses ant and needs to be able to build htsjdk*:  This is a problem, but we should really switch picard to get htsjdk through maven anyway since the current build system of manual checkout is very hacky.   We could upgrade picard to use gradle as well which would make it all very seamless.  
- *We'd break people's existing intellij projects:* This is a real pain, which is why I'd recommend switching to gradle as part of the move to java 8 compatible 2.0.0 instead of as a different release.  That way instead of requiring people to edit their project configs to switch to java 8, we can just have the reimport with gradle.
- *Rearranging the project structure is disruptive:*.  We'd want to switch from the ant project layout to the slightly different gradle default project layout.  I think this is a 1 time cost that isn't so bad, and the gradle layout is nicely symmetric in the end.  We would just have to watch pull requests to make sure they aren't introducing files in the old structure until they've all caught up.

Do people have thoughts, suggestions, or objections?  Is an issue the best way to discuss this?

@nh13 @tfenne @droazen @yfarjoun

**Comment by vadimzalunin** [comment]
i think this is a great idea

**Comment by tfenne** [comment]
Thanks for writing this up @lbergelson.  I agree that the gradle built script looks quite nice.  I'm curious if you looked at other builds tools also before settling on gradle for GATK 4?

Regarding the downsides you've listed, I'll comment on two of them:
* Rearranging is disruptive - it's not so bad.  I agree it's a bit of a pain in the butt to have to rename src directories and move every file in the repository to satisfy a build system, but so long as it's handled well it should be ok.  That said, does gradle not allow for overriding some of these things?
* The bigger one is the Picard -> htsjdk builds.  While using maven or similar dependency management is ok for release builds, a lot of of the time there are coordinated changes being made between HTSJDK and Picard (because we push so many utility classes down into HTSJDK).  I think we'd need to maintain a mechanism (doesn't have to be the default) where Picard can also build and use a local copy of HTSJDK.  I suspect this can be accomplished just fine even with HTSJDK being built by gradle.

The other thing I'd suggest bundling in with this change is removing the project files from source control.  Having IntelliJ project files in GitHub is worse than useless since IntelliJ constantly modifies them.  If gradle can generate the files, then there's no reason at all to have them in source control.

**Comment by nh13** [comment]
@lbergelson I have no objects.  I checked out your branch and found that the javadoc location has changed, as well as the htsjdk jar.  But we can go into those debates upon a PR.

**Comment by lbergelson** [comment]
@tfenne 
- Adam looked into build systems last year for hellbender and picked out gradle. I think the big requirement was no xml.  The obvious other choices were maven, which violates no xml in a big way, and sbt.  Sbt has the disadvantage that it's scala based, which is a farther jump from java than groovy is.  It also requires a scala installation as far as I know, which can be off putting for a casual downloader.  There are some other options like leiningen, but gradle seems to what a lot of java development is moving towards (at least as far as I know.. which is not that far outside of 75Ames)

- Gradle does allow you to customize the structure, but it seems like the pain of dealing with a non-standard structure forever is probably worse than the pain of moving all the files once.  I tried out the move and rebases of existing pull requests on  it go smoothly, git can track it all.  (except for new files, those end up in the wrong place without manual intervention).  I think the gradle structure is actually nicer in a lot of ways, it's symmetric between code and tests so it's always clear where things go.  
```
src -| main -| java
     |       | resources
     |
     | test -| java 
             | resources
```

- htsjdk -> picard integration is a bit of a pain.  The easiest thing to do would be to just keep the same process as now but replace ant with gradle.  
We publish snapshot builds with every commit to master (and could change it to be on every commit instead), so it's not to hard to co-develop them.  Just point to the snapshot for development and then update to the numbered version on release.  The cycle of push -> travis -> snapshot  is fairly slow for development iteration though.  We can either provide an option to build from a local htsjdk jar, or install htsjdk to a the local maven repo and pull it from there for development.  I'm sure there are other options as well. 

@nh13 My branch is definitely not ready for primetime.  I need input on it to make sure it has all the features that are needed.  For instance, I wasn't sure if we needed to be able to provide separate a separate tribble jar, and I noticed I'm not excluding some tests that ant excludes.  If people agree with the idea I'll open the pull request and people can have at it.

**Comment by lbergelson** [comment]
Forgot to say, but I strongly support removing the intellij files from source control.  They cause all sorts of problems for me.

**Comment by droazen** [comment]
A massive, enthusiastic :+1: to removing the checked-in IntelliJ project files -- we are constantly having to revert changes to those files before every commit. Gradle makes it easy/automatic to create an IntelliJ project, so the checked-in project files should no longer be needed with this change.

**Comment by jacarey** [comment]
This is great @lbergelson. :+1: I'd like to work out how to do the oss signed releases using gradle. This would be a good time to hand the public releases off to @ktibbett and the ops team.

**Comment by lbergelson** [comment]
Closed by #383.
